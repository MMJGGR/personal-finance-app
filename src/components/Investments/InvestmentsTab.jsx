import React, { useMemo } from 'react'
import { useFinance } from '../../FinanceContext'
import { formatCurrency } from '../../utils/formatters'
import AssetRow from './AssetRow'


  
  export default function InvestmentsTab() {
  const { assetsList, setAssetsList, settings } = useFinance()


  const handleAssetChange = (id, field, value) => {
    setAssetsList(prev =>
      prev.map(asset => (asset.id === id ? { ...asset, [field]: value } : asset))
    )
  }

  const addAsset = () => {
    setAssetsList(prev => [
      ...prev,
      {
        id: crypto.randomUUID(),
        name: '',
        amount: 0,
        type: 'Cash',
        expectedReturn: 0,
        volatility: 0,
        horizonYears: 0,
        purchaseYear: new Date().getFullYear(),
        saleYear: null,
        principal: 0,
      },
    ])
  }

  const removeAsset = id => {
    if (window.confirm('Delete this asset?')) {
      setAssetsList(prev => prev.filter(asset => asset.id !== id))
    }
  }

  const totalAssets = useMemo(() => {
    return assetsList.reduce((sum, asset) => sum + (Number(asset.amount) || 0), 0)
  }, [assetsList])

  return (
    <div className="space-y-8 p-6">
      <h2 className="text-2xl font-bold text-amber-800">Investments</h2>

      <section>
        <h3 className="text-xl font-bold text-amber-800 mb-4">Asset List</h3>
        <div className="grid gap-4">
          {assetsList.length === 0 && (
            <p className="italic text-slate-500 col-span-full text-center">No assets added yet</p>
          )}
          {assetsList.map(asset => (
            <AssetRow
              key={asset.id}
              asset={asset}
              onChange={handleAssetChange}
              onDelete={removeAsset}
              currency={settings.currency}
            />
          ))}
        </div>
        <button
          onClick={addAsset}
          className="mt-4 bg-amber-400 hover:bg-amber-300 text-white px-4 py-2 rounded-md focus:outline-none focus:ring-2 focus:ring-amber-500"
        >
          âž• Add Asset
        </button>
      </section>

      <section className="bg-white p-4 rounded-xl shadow-md">
        <h3 className="text-lg font-bold text-amber-800 mb-2">Portfolio Summary</h3>
        <p>Total Assets: {formatCurrency(totalAssets, settings.locale, settings.currency)}</p>
        {/* Add more portfolio overview details here, e.g., allocation pie chart */}
      </section>
    </div>
  )
}
