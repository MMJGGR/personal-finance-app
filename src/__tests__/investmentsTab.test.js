import React from 'react'
import { render } from '@testing-library/react'
import { FinanceProvider } from '../FinanceContext'
import InvestmentsTab from '../components/Investments/InvestmentsTab'

beforeAll(() => {
  global.ResizeObserver = class { observe() {} unobserve() {} disconnect() {} }
})

test('investments tab placeholder snapshot', () => {
  const { container } = render(
    <FinanceProvider>
      <InvestmentsTab />
    </FinanceProvider>
  )
  expect(container.firstChild).toMatchSnapshot()
})
