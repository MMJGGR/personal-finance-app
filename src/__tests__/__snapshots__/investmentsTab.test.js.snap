// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`investments tab placeholder snapshot 1`] = `
<div
  class="space-y-8 p-6"
>
  <h2
    class="text-2xl font-bold text-amber-800"
  >
    Investments
  </h2>
  <section>
    <h3
      class="text-xl font-bold text-amber-800 mb-4"
    >
      Asset List
    </h3>
    <div
      class="grid gap-4"
    >
      <div
        class="grid grid-cols-1 sm:grid-cols-7 gap-2 items-center mb-1 bg-white p-2 rounded-md shadow relative"
      >
        <div>
          <label
            class="block text-sm font-medium"
            for="fc853948-cc17-4c84-ab8b-7a71cb2e6e73-name"
          >
            Name
          </label>
          <input
            class="border p-2 rounded-md w-full"
            id="fc853948-cc17-4c84-ab8b-7a71cb2e6e73-name"
            type="text"
            value="Cash"
          />
        </div>
        <div>
          <label
            class="block text-sm font-medium"
            for="fc853948-cc17-4c84-ab8b-7a71cb2e6e73-amount"
          >
            Amount (
            KES
            )
          </label>
          <input
            class="border p-2 rounded-md w-full text-right"
            id="fc853948-cc17-4c84-ab8b-7a71cb2e6e73-amount"
            type="number"
            value="500000"
          />
        </div>
        <div>
          <label
            class="block text-sm font-medium"
            for="fc853948-cc17-4c84-ab8b-7a71cb2e6e73-type"
          >
            Type
          </label>
          <select
            class="border p-2 rounded-md w-full"
            id="fc853948-cc17-4c84-ab8b-7a71cb2e6e73-type"
          >
            <option
              value="Cash"
            >
              Cash
            </option>
            <option
              value="Portfolio"
            >
              Portfolio
            </option>
            <option
              value="Real Estate"
            >
              Real Estate
            </option>
            <option
              value="Other"
            >
              Other
            </option>
          </select>
        </div>
        <div>
          <label
            class="block text-sm font-medium"
            for="fc853948-cc17-4c84-ab8b-7a71cb2e6e73-expectedReturn"
          >
            Expected Return (%)
          </label>
          <input
            class="border p-2 rounded-md w-full text-right"
            id="fc853948-cc17-4c84-ab8b-7a71cb2e6e73-expectedReturn"
            step="0.1"
            type="number"
            value="2"
          />
        </div>
        <div>
          <label
            class="block text-sm font-medium"
            for="fc853948-cc17-4c84-ab8b-7a71cb2e6e73-volatility"
          >
            Volatility (%)
          </label>
          <input
            class="border p-2 rounded-md w-full text-right"
            id="fc853948-cc17-4c84-ab8b-7a71cb2e6e73-volatility"
            step="0.1"
            type="number"
            value="1"
          />
        </div>
        <div>
          <label
            class="block text-sm font-medium"
            for="fc853948-cc17-4c84-ab8b-7a71cb2e6e73-horizonYears"
          >
            Horizon (Years)
          </label>
          <input
            class="border p-2 rounded-md w-full text-right"
            id="fc853948-cc17-4c84-ab8b-7a71cb2e6e73-horizonYears"
            step="1"
            type="number"
            value="0"
          />
        </div>
        <button
          class="absolute top-1 right-1 text-lg"
        >
          ❌
        </button>
      </div>
      <div
        class="grid grid-cols-1 sm:grid-cols-7 gap-2 items-center mb-1 bg-white p-2 rounded-md shadow relative"
      >
        <div>
          <label
            class="block text-sm font-medium"
            for="5b218e09-7e00-41f9-8430-e34cca70cf68-name"
          >
            Name
          </label>
          <input
            class="border p-2 rounded-md w-full"
            id="5b218e09-7e00-41f9-8430-e34cca70cf68-name"
            type="text"
            value="Investments"
          />
        </div>
        <div>
          <label
            class="block text-sm font-medium"
            for="5b218e09-7e00-41f9-8430-e34cca70cf68-amount"
          >
            Amount (
            KES
            )
          </label>
          <input
            class="border p-2 rounded-md w-full text-right"
            id="5b218e09-7e00-41f9-8430-e34cca70cf68-amount"
            type="number"
            value="1000000"
          />
        </div>
        <div>
          <label
            class="block text-sm font-medium"
            for="5b218e09-7e00-41f9-8430-e34cca70cf68-type"
          >
            Type
          </label>
          <select
            class="border p-2 rounded-md w-full"
            id="5b218e09-7e00-41f9-8430-e34cca70cf68-type"
          >
            <option
              value="Cash"
            >
              Cash
            </option>
            <option
              value="Portfolio"
            >
              Portfolio
            </option>
            <option
              value="Real Estate"
            >
              Real Estate
            </option>
            <option
              value="Other"
            >
              Other
            </option>
          </select>
        </div>
        <div>
          <label
            class="block text-sm font-medium"
            for="5b218e09-7e00-41f9-8430-e34cca70cf68-expectedReturn"
          >
            Expected Return (%)
          </label>
          <input
            class="border p-2 rounded-md w-full text-right"
            id="5b218e09-7e00-41f9-8430-e34cca70cf68-expectedReturn"
            step="0.1"
            type="number"
            value="8"
          />
        </div>
        <div>
          <label
            class="block text-sm font-medium"
            for="5b218e09-7e00-41f9-8430-e34cca70cf68-volatility"
          >
            Volatility (%)
          </label>
          <input
            class="border p-2 rounded-md w-full text-right"
            id="5b218e09-7e00-41f9-8430-e34cca70cf68-volatility"
            step="0.1"
            type="number"
            value="15"
          />
        </div>
        <div>
          <label
            class="block text-sm font-medium"
            for="5b218e09-7e00-41f9-8430-e34cca70cf68-horizonYears"
          >
            Horizon (Years)
          </label>
          <input
            class="border p-2 rounded-md w-full text-right"
            id="5b218e09-7e00-41f9-8430-e34cca70cf68-horizonYears"
            step="1"
            type="number"
            value="5"
          />
        </div>
        <button
          class="absolute top-1 right-1 text-lg"
        >
          ❌
        </button>
      </div>
      <div
        class="grid grid-cols-1 sm:grid-cols-7 gap-2 items-center mb-1 bg-white p-2 rounded-md shadow relative"
      >
        <div>
          <label
            class="block text-sm font-medium"
            for="pv-income-name"
          >
            Name
          </label>
          <input
            class="border p-2 rounded-md w-full"
            id="pv-income-name"
            type="text"
            value="PV of Lifetime Income"
          />
        </div>
        <div>
          <label
            class="block text-sm font-medium"
            for="pv-income-amount"
          >
            Amount (
            KES
            )
          </label>
          <input
            class="border p-2 rounded-md w-full text-right"
            id="pv-income-amount"
            type="number"
            value="NaN"
          />
        </div>
        <div>
          <label
            class="block text-sm font-medium"
            for="pv-income-type"
          >
            Type
          </label>
          <select
            class="border p-2 rounded-md w-full"
            id="pv-income-type"
          >
            <option
              value="Cash"
            >
              Cash
            </option>
            <option
              value="Portfolio"
            >
              Portfolio
            </option>
            <option
              value="Real Estate"
            >
              Real Estate
            </option>
            <option
              value="Other"
            >
              Other
            </option>
          </select>
        </div>
        <div>
          <label
            class="block text-sm font-medium"
            for="pv-income-expectedReturn"
          >
            Expected Return (%)
          </label>
          <input
            class="border p-2 rounded-md w-full text-right"
            id="pv-income-expectedReturn"
            step="0.1"
            type="number"
            value="0"
          />
        </div>
        <div>
          <label
            class="block text-sm font-medium"
            for="pv-income-volatility"
          >
            Volatility (%)
          </label>
          <input
            class="border p-2 rounded-md w-full text-right"
            id="pv-income-volatility"
            step="0.1"
            type="number"
            value="0"
          />
        </div>
        <div>
          <label
            class="block text-sm font-medium"
            for="pv-income-horizonYears"
          >
            Horizon (Years)
          </label>
          <input
            class="border p-2 rounded-md w-full text-right"
            id="pv-income-horizonYears"
            step="1"
            type="number"
            value="0"
          />
        </div>
        <button
          class="absolute top-1 right-1 text-lg"
        >
          ❌
        </button>
      </div>
      <div
        class="grid grid-cols-1 sm:grid-cols-7 gap-2 items-center mb-1 bg-white p-2 rounded-md shadow relative"
      >
        <div>
          <label
            class="block text-sm font-medium"
            for="projected-pension-value-name"
          >
            Name
          </label>
          <input
            class="border p-2 rounded-md w-full"
            id="projected-pension-value-name"
            type="text"
            value="Projected Pension Value"
          />
        </div>
        <div>
          <label
            class="block text-sm font-medium"
            for="projected-pension-value-amount"
          >
            Amount (
            KES
            )
          </label>
          <input
            class="border p-2 rounded-md w-full text-right"
            id="projected-pension-value-amount"
            type="number"
            value="0"
          />
        </div>
        <div>
          <label
            class="block text-sm font-medium"
            for="projected-pension-value-type"
          >
            Type
          </label>
          <select
            class="border p-2 rounded-md w-full"
            id="projected-pension-value-type"
          >
            <option
              value="Cash"
            >
              Cash
            </option>
            <option
              value="Portfolio"
            >
              Portfolio
            </option>
            <option
              value="Real Estate"
            >
              Real Estate
            </option>
            <option
              value="Other"
            >
              Other
            </option>
          </select>
        </div>
        <div>
          <label
            class="block text-sm font-medium"
            for="projected-pension-value-expectedReturn"
          >
            Expected Return (%)
          </label>
          <input
            class="border p-2 rounded-md w-full text-right"
            id="projected-pension-value-expectedReturn"
            step="0.1"
            type="number"
            value="8"
          />
        </div>
        <div>
          <label
            class="block text-sm font-medium"
            for="projected-pension-value-volatility"
          >
            Volatility (%)
          </label>
          <input
            class="border p-2 rounded-md w-full text-right"
            id="projected-pension-value-volatility"
            step="0.1"
            type="number"
            value="0"
          />
        </div>
        <div>
          <label
            class="block text-sm font-medium"
            for="projected-pension-value-horizonYears"
          >
            Horizon (Years)
          </label>
          <input
            class="border p-2 rounded-md w-full text-right"
            id="projected-pension-value-horizonYears"
            step="1"
            type="number"
            value="25"
          />
        </div>
        <button
          class="absolute top-1 right-1 text-lg"
        >
          ❌
        </button>
      </div>
    </div>
    <button
      class="mt-4 bg-amber-400 hover:bg-amber-300 text-white px-4 py-2 rounded-md focus:outline-none focus:ring-2 focus:ring-amber-500"
    >
      ➕ Add Asset
    </button>
  </section>
  <section
    class="bg-white p-4 rounded-xl shadow-md"
  >
    <h3
      class="text-lg font-bold text-amber-800 mb-2"
    >
      Portfolio Summary
    </h3>
    <p>
      Total Assets: 
      1.5 M
    </p>
  </section>
</div>
`;
